FILE_SUFFIX fixes

57300 opened the pull request:
* Fix doubling `FILE_SUFFIX` on `boards/*.overlay files`
* Use default `FILE_SUFFIX` in `zephyr_file()` (wasn't working in sysbuild)
* Bonus: use `FILE_SUFFIX` to simplify a sample's configuration

57300 committed:
cmake: extensions: Change zephyr_file() argument parser prefix

Replace `FILE` with `ZFILE` as the prefix for `cmake_parse_arguments()`.
This is because `zephyr_file()` accepts `SUFFIX` as function argument,
and it should not be accessible as `FILE_SUFFIX` as it would conflict
with the CMake cache variable of the same name. Accessing the argument
as `ZFILE_SUFFIX` prevents the cached value from slipping into the
function scope.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>

57300 committed:
cmake: configuration_files: Add missing FILE_SUFFIX handling

for Kconfig fragments in the `boards` application subdirectory.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>

57300 committed:
samples: drivers: mbox: Use FILE_SUFFIX

Simplify this sample's recently updated configuration, and ensure that
the file suffix feature works in sysbuild.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>

57300 committed:
cmake: Fix doubling FILE_SUFFIX on boards/*.overlay files

When building with `-DBOARD=plank -DFILE_SUFFIX=foo`, the build system
could set `DTC_OVERLAY_FILE` to one of:

  - boards/plank_foo_foo.overlay
  - boards/plank_foo.overlay
  - boards/plank.overlay

The first choice is a bug, as the suffix should only be applied once.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>

nordicjm reviewed (commented):
Seems it is not working in sysbuild, share/sysbuild/cmake/modules/sysbuild_kconfig.cmake
```
if(DEFINED SB_CONF_FILE)
  # SB_CONF_FILE already set so nothing to do.
elseif(DEFINED ENV{SB_CONF_FILE})
  set(SB_CONF_FILE $ENV{SB_CONF_FILE})
elseif(EXISTS      ${APP_DIR}/sysbuild.conf)
  set(SB_CONF_FILE ${APP_DIR}/sysbuild.conf)
else()
  # Because SYSBuild is opt-in feature, then it is permitted to not have a
  # SYSBuild dedicated configuration file.
endif()
```
would need to be updated to support it

nordicjm commented:
Just noticed another issue, `-DSB_FILE_SUFFIX` does not work for `<app_dir>/sysbuild/<app>/prj_<suffix>.conf`

57300 committed:
sysbuild: Apply FILE_SUFFIX to sysbuild.conf

Unless `SB_CONF_FILE` is set by the user, an optional `sysbuild.conf`
file is taken by default. If `FILE_SUFFIX=foo` is set, then it should
also be possible to find `sysbuild_foo.conf`, just like with `prj.conf`.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>

57300 commented:
> Seems it is not working in sysbuild, share/sysbuild/cmake/modules/sysbuild_kconfig.cmake

Right, this wasn't implemented for `SB_CONF_FILE`. Added another commit.

> Just noticed another issue, `-DSB_FILE_SUFFIX` does not work for `<app_dir>/sysbuild/<app>/prj_<suffix>.conf`

There is no `SB_FILE_SUFFIX`. The regular `FILE_SUFFIX` should work here.

nordicjm commented:
> > Maybe I'm mixing it up with sysbuild. If you use `-Dsysbuild_FILE_SUFFIX=x` can you check it works?
> 
> `-D<image>_FILE_SUFFIX` works. `sysbuild` is not a valid prefix afaik.

We should have one for sysbuild @tejlmand

tejlmand commented on cmake/modules/extensions.cmake:
```diff
@@ -2594,6 +2594,10 @@ Relative paths are only allowed with `-D${ARGV1}=<path>`")
         set(ZFILE_BOARD_IDENTIFIER ${BOARD_IDENTIFIER})
       endif()
     endif()
+    if(NOT DEFINED ZFILE_SUFFIX)
+      # Defaulting to system wide settings when SUFFIX is not given as argument
+      set(ZFILE_SUFFIX ${FILE_SUFFIX})
+    endif()
```
this change impact behavior for cases where suffix should not be applied, such as board and shield dir files.
https://github.com/zephyrproject-rtos/zephyr/blob/9a0b8e7809b0e877d2be664dcc09f9ba117b8a19/cmake/modules/dts.cmake#L157
https://github.com/zephyrproject-rtos/zephyr/blob/9a0b8e7809b0e877d2be664dcc09f9ba117b8a19/cmake/modules/shields.cmake#L118
https://github.com/zephyrproject-rtos/zephyr/blob/9a0b8e7809b0e877d2be664dcc09f9ba117b8a19/cmake/modules/shields.cmake#L122

57300 replied:
This behavior was already present due to `FILE_SUFFIX` being inherited from cache scope.

If we don't want to apply the suffix in these places, then I can fix it now by dropping one of the commits and instead adding explicit `SUFFIX ${FILE_SUFFIX}` where it's needed for sure.

tejlmand replied:
I think we should avoid having FILE_SUFFIX applied on boards as that will open for misuse.
Instead hwmv2 provides board variant options.

nordicjm replied:
Agreed, files in `<board_dir>/<vendor>/<board>` should not have this applied, it should be applied to `<app>/<boards>/<board>.overlay` and `<app>/<boards>/<board>.dts` files

57300 replied:
Done.

tejlmand reviewed (changes requested):
we need to decide on how to use `FILE_SUFFIX` per default.

As now, it's an application feature so that application may have different configuration depending on use-case.

File suffix'es should not be applied to board / shield dirs under board_root, as those has variants and revisions.

Thus defaulting to pickup suffix'es is not ideal.

57300 commented on cmake/modules/extensions.cmake:
```diff
@@ -2535,75 +2535,75 @@ Please provide one of following: APPLICATION_ROOT, CONF_FILES")
-  cmake_parse_arguments(FILE "${options}" "${single_args}" "${multi_args}" ${ARGN})
-  if(FILE_UNPARSED_ARGUMENTS)
-      message(FATAL_ERROR "zephyr_file(${ARGV0} <val> ...) given unknown arguments: ${FILE_UNPARSED_ARGUMENTS}")
+  cmake_parse_arguments(ZFILE "${options}" "${single_args}" "${multi_args}" ${ARGN})
```
> When I first looked at the commit it wasn't clear to me that this was actually preparation work to allow picking up `FILE_SUFFIX` from parent scope.

Not exactly. As a standalone change, it has additional, unrelated effects:
* Prevent `FILE_SUFFIX` from being (incorrectly) inherited from cache scope
* Support `zephyr_boilerplate_watch(FILE_SUFFIX)` if we ever decide we need it

I'll try to rephrase the commit message though.

57300 replied:
> > then any parent or cached scoped FILE_SUFFIX is no longer available, even when SUFFIX is not passed as argument to zephyr_file().
> 
> Oh my, it does seem like cache variables can slip through. Guess a CMake issue should be raised.

Is that unexpected? I figured that if `SUFFIX` was not passed, then `cmake_parse_arguments()` would set `FILE_SUFFIX` to undefined in local scope, which is why the cached value slipped through.

tejlmand replied:
https://gitlab.kitware.com/cmake/cmake/-/issues/25773

tejlmand replied:
`ZFILE` is fine for now, but seems CMake might be opting for having a recommendation to use something like `ARG` / `arg` (not sure i'm fond of that, but let's wait and see)

Ref: https://gitlab.kitware.com/cmake/cmake/-/issues/25773#note_1499098

57300 commented on cmake/modules/configuration_files.cmake:
```diff
@@ -73,9 +73,10 @@ To change CONF_FILE, use the CONF_FILE variable." ${CONF_FILE_FORCE_CACHE})
 zephyr_get(DTC_OVERLAY_FILE SYSBUILD LOCAL)
+if(NOT DEFINED DTC_OVERLAY_FILE)
+  zephyr_file(CONF_FILES ${APPLICATION_CONFIG_DIR}/boards DTS DTC_OVERLAY_FILE SUFFIX ${FILE_SUFFIX})
+endif()
```
Added comments.

nordicjm reviewed (commented):
Problem I have with this is it seems impossible to use a file suffix with just sysbuild e.g. `-Dsysbuild_FILE_SUFFIX=<x>`

nordicjm commented on share/sysbuild/cmake/modules/sysbuild_kconfig.cmake:
```diff
@@ -17,11 +17,9 @@ if(DEFINED SB_CONF_FILE)
 else()
-  # Because SYSBuild is opt-in feature, then it is permitted to not have a
-  # SYSBuild dedicated configuration file.
+  # sysbuild.conf is an optional file, because sysbuild is an opt-in feature.
+  zephyr_file(CONF_FILES ${APP_DIR} KCONF SB_CONF_FILE NAMES "sysbuild.conf")
```
Shouldn't this have `SUFFIX` as part of it?

57300 replied:
Yes, thanks for spotting

57300 commented:
> Problem I have with this is it seems impossible to use a file suffix with just sysbuild e.g. `-Dsysbuild_FILE_SUFFIX=<x>`

It's not a bug and I don't know why we need it.
